service/storage_proxy.cc

```diff
@@ -1750,6 +1750,31 @@ storage_proxy::create_write_response_handler_helper(schema_ptr s, const dht::tok
     std::vector<gms::inet_address> natural_endpoints = rs.get_natural_endpoints(token);
     std::vector<gms::inet_address> pending_endpoints = _token_metadata.pending_endpoints_for(token, keyspace_name);
 
+    if (get_local_storage_service().get_token_metadata().is_any_node_being_replaced() &&
+        rs.get_type() != locator::replication_strategy_type::local) {
+        // When a new node is started to replace an existing dead node, we want
+        // to make the replacing node take writes but do not count it for
+        // consistency level, because the replacing node can die and go away.
+        // To do this, we filter out the existing node being replaced from
+        // natural_endpoints and make the replacing node in the pending_endpoints.
+        //
+        // However, we can not apply the filter for LocalStrategy because
+        // LocalStrategy always has RF = 1 and returns the node itself as the
+        // natural_endpoints. If the replacing node had the same ip address as
+        // the node being replaced, we can not filter out the local node,
+        // otherwise all the query to keyspace with LocalStrategy will fail,
+        // e.g., the system keyspace.
+        auto original_natural_endpoints = natural_endpoints;
+        auto it = boost::range::remove_if(natural_endpoints, [] (gms::inet_address& p) {
+            return get_local_storage_service().get_token_metadata().is_being_replaced(p);
+        });
```

Comment:
What do you mean by confused with ip? What do you suggest we do?  What do you disagree with this approach?

Here is what Gleb and I discussed and agreed on:

https://github.com/scylladb/scylla/issues/5482#issuecomment-576165303

Reply:
I mentioned this problem in the previous comment: https://github.com/scylladb/scylla/pull/5731#discussion_r403124032

"Node being confused with ip" refers to the fact that we have two different nodes here, replaced and replacing, but both are identified in the system as one thing because they share the IP. That's why you need to filter natural endpoints, assuming that the IP in the natural endpoints refers to the replaced node, and assume that the IP in the pending endpoints refers to the replacing node. If those two nodes wouldn't share the identified, you wouldn't have to filter natural endpoints.

Reply:
It would be better if we didn't identify nodes by IP, but something else which uniquely identifies them:

```
class endpoint {
    class impl;
    lw_shared_ptr<impl> _impl;
public:
    gms::inet_address ip() const;
    UUID host_id() const;
    bool is_up() const;
    bool operator==(const endpoint&) const;
};
```

Could we use hostID to distinguish replacing from replaced?

Reply:
No. Replacing node uses the same host id and tokens of the replaced node. Replacing node could use a different IP address or same IP address.